File: src/c4/inter/Else.java
package inter;

import treewalker.TreeWalker;

/*
 * Else ist eine Unterklasse von Stmt. In der Instanzenvariable expr
 * wird der Ausdruck, in stmt1 und stmt2 die beiden Anweisungen abgelegt.
 */

public class Else extends Stmt {
	Expr expr;
	Stmt stmt1, stmt2;

	public Else(Expr x, Stmt s1, Stmt s2) {
		expr = x;
		stmt1 = s1;
		stmt2 = s2;
	}

	public Expr getExpr() {
		return expr;
	}
	
	public void setExpr(Expr expr) {
		this.expr = expr;
	}


	public Stmt getStmt1() {
		return stmt1;
	}

	public Stmt getStmt2() {
		return stmt2;
	}

	public <R, P> R walk(TreeWalker<R, P> walker, P arg) {
		return walker.walkElseNode(this, arg);
	}

}

File: src/c4/inter/Else.rs
use super::{Expr, Node, NodeData, StatementNode, Stmt, StmtData, TreeWalker, Walk};

#[derive(Clone, Debug, PartialEq)]
pub struct Else {
    data: StmtData,
    expr: Expr,
    stmt1: Stmt,
    stmt2: Stmt,
}

impl Else {
    pub fn new(expr: Expr, stmt1: Stmt, stmt2: Stmt) -> Self {
        Self {
            data: StmtData::new(),
            expr,
            stmt1,
            stmt2,
        }
    }

    pub fn get_expr(&self) -> &Expr {
        &self.expr
    }
    pub fn set_expr(&mut self, expr: Expr) {
        self.expr = expr;
    }
    pub fn get_stmt1(&self) -> &Stmt {
        &self.stmt1
    }
    pub fn get_stmt2(&self) -> &Stmt {
        &self.stmt2
    }
}

impl Node for Else {
    fn node_data(&self) -> &NodeData {
        &self.data.node
    }
}

impl StatementNode for Else {
    fn stmt_data(&self) -> &StmtData {
        &self.data
    }
    fn stmt_data_mut(&mut self) -> &mut StmtData {
        &mut self.data
    }
}

impl<P, R> Walk<P, R> for Else {
    fn walk<W: TreeWalker<P, R> + ?Sized>(&self, walker: &mut W, arg: P) -> R {
        walker.walk_else_node(self, arg)
    }
}
